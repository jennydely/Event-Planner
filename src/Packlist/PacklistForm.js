import React, { useState, useRef } from 'react'
import { useForm } from 'react-hook-form'
import { v4 as uuid } from 'uuid'
import styled from 'styled-components/macro'
import Input from '../common/Input'
import Label from '../common/Label'
import ErrorMessage from '../common/ErrorMessage'
import PropTypes from 'prop-types'
import ListContainer from '../common/ListContainer'
import ListItem from '../common/ListItem'

PacklistForm.propTypes = {
  onPacklistSave: PropTypes.func.isRequired,
}

export default function PacklistForm({onPacklistSave }) {
  const { register, handleSubmit, reset, errors } = useForm()
  const [items, setItems] = useState([])
  const ItemRef = useRef(null)
  const onSubmit = (packlist, event) => {
    // for testing...
    if (event && event.target && typeof event.target.reset === 'function')

      event.target.reset()
    onPacklistSave({ name: packlist.name, packlist: [packlist.item] })
  }



  return (
    <>
      <h1>Create PackList</h1>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <FormInputContainer>
          <PacklistNameInputLabel>Name of the new PackList:</PacklistNameInputLabel>
          <PacklistNameInput
            placeholder="PackList name"
            id="name"
            name="name"
            ref={register({
              required: true, minLength: 3, maxLength: 10,
              validate: value => value && value.trim().length >= 3 && value.trim().length <= 10
            })} />
          {errors.name && errors.name.type === 'required' && (
            <ErrorMessageNameReq>Name is required!</ErrorMessageNameReq>
          )}
          {errors.name && (errors.name.type === 'validate' || errors.name.type === 'minLength') && (
            <ErrorMessageName>
              This field requires at least 3 characters!
            </ErrorMessageName>
          )}
          {errors.name && (errors.name.type === 'validate' || errors.name.type === 'maxLength') && (
            <ErrorMessageName>
              The name can reach a maximum of 10 characters!
            </ErrorMessageName>
          )}
          <Label htmlFor="itemInput">Create new item or task:</Label>
          <Input
            placeholder="item you need or task you have to do"
            id="itemInput"
            name="item"
            ref={(el) => {
              ItemRef.current = el
              register(el, {
                required: true, minLength: 3, maxLength: 20,
                validate: value => value && value.trim().length >= 3 && value.trim().length <= 20
              })
            }}
          />
          <AddButton type="button" onClick={createItem}>Add</AddButton>

          {errors.item && (errors.item.type === 'validate' || errors.item.type === 'minLength') && (
            <ErrorMessageNameReq>
              This field requires at least 3 characters!
            </ErrorMessageNameReq>
          )}
          {errors.item && (errors.item.type === 'validate' || errors.item.type === 'maxLength') && (
            <ErrorMessage>
              The Item can reach a maximum of 20 characters!
            </ErrorMessage>
          )}
          <ItemContainer>
            {items.map(({ text, completed, id }, index) => (
              <ListItem key={id} text={text}><input type="checkbox" checked={completed} />{text}  <DeleteButton onClick={() => deleteItem(index)} type="button">X</DeleteButton> </ListItem>
            ))}
          </ItemContainer>
        </FormInputContainer>

        <ButtonGroup>
          <button type="reset" onClick={() => reset()}>
            Reset
          </button>
          <SubmitButton type="submit">Save</SubmitButton>
        </ButtonGroup>
      </Form>
    </>

  )
  function createItem(itemRef) {
    itemRef.current.value.preventDefault()
    console.log(itemRef.current.value)
    addItem({ text: itemRef.current.value, id: uuid() })
  }

  function addItem(item) {
    setItems([item, ...items])
  }

  function deleteItem(index) {
    setItems([
      ...items.slice(0, index),
      ...items.slice(index + 1)
    ])
  }
}

const Form = styled.form`
align-content: center;
min-width: 300px;
`
const FormInputContainer = styled.div`
display:flex;
flex-direction: column;
align-items: flex-start;
min-width: 300px;
border-radius: 7px;
border: 2px solid black;
margin: 7px;
margin-top:0;
padding: 7px 4px;
padding-bottom: 20px;
`
const PacklistNameInputLabel = styled(Label)`
text-align:left;
`
const PacklistNameInput = styled(Input)`
`
const ErrorMessageNameReq = styled(ErrorMessage)`
`
const ErrorMessageName = styled(ErrorMessage)`
`
const AddButton = styled.button`
align-self:center;
`
const ItemContainer = styled(ListContainer)`
border: none;
`
const ButtonGroup = styled.div`
grid-column: 1;
grid-row: 8;
display: flex;
justify-content: center;
width: 100%;
maring:7px;
margin-top: 30px;
`
const SubmitButton = styled.button`
background-color: rgba(111,29,27,0.75);
`
const DeleteButton = styled.button`
color: rgba(246, 71, 71, 1);
text-decoration: none;
border:none;
background:none;
`