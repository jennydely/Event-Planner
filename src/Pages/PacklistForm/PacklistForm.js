import React, { useState, useRef, useEffect } from 'react'
import { useForm } from 'react-hook-form'
import { v4 as uuid } from 'uuid'
import { useParams, useHistory } from 'react-router-dom'
import styled from 'styled-components/macro'
import { connect } from 'react-redux'
import { editPacklist, addPacklist } from '../../redux/actions'
import Input from '../components/common/Input'
import Label from '../components/common/Label'
import Checkbox from '../components/common/Checkbox'
import ErrorMessage from '../components/common/ErrorMessage'
import SelectTemplate from '../components/common/SelectTemplate'
import ListContainer from '../components/common/ListContainer'
import ListItem from '../components/common/ListItem'
import FormHeader from '../components/FormHeader'
import Footer from '../components/FormFooter'
import addIcon from '../../img/addIcon.svg'
import usePacklists from '../Packlist/usePacklists'
import { getUniquePacklists } from './services/getUniquePacklists'
import { getUniqueItems } from './services/getUniqueItems'

export function PacklistForm({ editPacklist, addPacklist }) {
  const { register, handleSubmit, errors } = useForm()
  const { packlistId } = useParams()
  const { packlists } = usePacklists()
  const packlistToEdit = searchForEditPacklist()
  const [items, setItems] = useState([])
  const [itemError, setItemError] = useState(false)
  const itemRef = useRef(null)

  const history = useHistory()
  const uniquePacklists = getUniquePacklists(packlists)
  const uniqueItems = getUniqueItems(items)
  const onSubmit = (packlist, event) => {
    event.preventDefault()
    const isExistingPacklist = packlists.some(
      (existingPacklist) => packlist.name === existingPacklist.name
    )

    if (isExistingPacklist) {
      editPacklist({
        ...packlist,
        id: packlistToEdit.id,
        packlist: items,
      })
    } else {
      addPacklist({ name: packlist.name, packlist: items, id: uuid() })
    }
    history.push('/packlist/' + packlist.name)
  }
  const handlePacklistKeyDown = (event) => {
    if (event.key === 'Enter') {
      event.preventDefault()
      addItem({ item: event.target.value, itemID: uuid() })
    }
  }
  useEffect(() => {
    if (packlistToEdit?.packlist) setItems(packlistToEdit?.packlist)
  }, [packlistToEdit])

  return (
    <>
      <FormHeader
        headerText={packlistToEdit ? 'Edit PackList' : 'Create PackList'}
      />
      <main>
        <Form data-testid="packlistform" onSubmit={handleSubmit(onSubmit)}>
          <FormInputContainer>
            <PacklistNameInputLabel htmlFor="name">
              Name of the new PackList:
            </PacklistNameInputLabel>
            <Input
              placeholder="PackList name"
              id="name"
              name="name"
              defaultValue={packlistToEdit?.name}
              ref={register({
                required: true,
                minLength: 3,
                maxLength: 20,
                validate: {
                  length: (value) =>
                    value?.trim().length >= 3 && value?.trim().length <= 20,
                  nameTaken: (value) =>
                    !!packlistId ||
                    !packlists.some((packlist) => packlist.name === value),
                },
              })}
            />
            {errors.name?.type === 'required' && (
              <ErrorMessage>Name is required!</ErrorMessage>
            )}
            {errors.name?.type === 'nameTaken' && (
              <ErrorMessage>Name is taken!</ErrorMessage>
            )}
            {(errors.name?.type === 'validate' ||
              errors.name?.type === 'minLength') && (
                <ErrorMessage>
                  This field requires at least 3 characters!
                </ErrorMessage>
              )}
            {(errors.name?.type === 'validate' ||
              errors.name?.type === 'maxLength') && (
                <ErrorMessage>
                  The name can reach a maximum of 20 characters!
                </ErrorMessage>
              )}
            <Label htmlFor="itemInput">Create new item or task:</Label>
            <Input
              placeholder="Item you need or task you have to do"
              id="itemInput"
              name="item"
              onKeyDown={handlePacklistKeyDown}
              ref={(el) => {
                itemRef.current = el
              }}
            />
            <HiddenInput
              name="shouldHaveOneItem"
              ref={register({
                validate: () => {
                  return items.length > 0
                },
              })}
            />

            {errors.shouldHaveOneItem?.type === 'validate' && (
              <ErrorMessage>You need to add one item/task!</ErrorMessage>
            )}

            {itemError && (
              <ErrorMessage>
                'This item field requires at least 3 characters and can reach a
                maximum of 33 characters!'
              </ErrorMessage>
            )}

            <AddButton
              type="button"
              onClick={() => {
                addItem({ item: itemRef.current.value, itemID: uuid() })
              }}
            >
              <img src={addIcon} alt="add" />
            </AddButton>
            <StyledListContainer>
              {uniqueItems?.map(({ item, completed, itemID }, index) => (
                <ListItem key={item} id={itemID} text={item}>
                  <ItemContainer>
                    <Checkbox type="checkbox" checked={completed} />
                    <TextSpan>{item}</TextSpan>
                  </ItemContainer>
                  <DeleteButton onClick={() => deleteItem(index)} type="button">
                    X
                  </DeleteButton>
                </ListItem>
              ))}
            </StyledListContainer>
            <Label htmlFor="packlistTemplate">Use packlist as template:</Label>
            <TemplateDropDown
              id="packlistTemplate"
              defaultValue=" "
              name="packlistName"
              options={uniquePacklists}
              addMultiplyItems={addMultiplyItems}
            />
          </FormInputContainer>
        </Form>
      </main>
      <Footer handleSubmit={handleSubmit(onSubmit)} />
    </>
  )

  function addItem(item) {
    if (item.item?.trim().length >= 3 && item.item?.trim().length <= 25) {
      setItems([item, ...items])
      itemRef.current.value = ''
      setItemError(false)
    } else setItemError(true)
  }

  function addMultiplyItems(templateName) {
    const templatePacklist = packlists.find(
      (packlist) => templateName.name === packlist.name
    )
    setItems([...templatePacklist.packlist, ...items])
  }

  function deleteItem(index) {
    setItems([...items.slice(0, index), ...items.slice(index + 1)])
  }

  function searchForEditPacklist() {
    const index = packlists.findIndex(
      (packlist) => '' + packlist.id === packlistId
    )
    const editPacklist = packlists[index]
    return editPacklist
  }
}

const mapDispatchToProps = { addPacklist, editPacklist }

export default connect(
  null, 
  mapDispatchToProps
  )  (PacklistForm)

const Form = styled.form`
  align-content: center;
  overflow: hidden;
  min-width: 365px;
`
const FormInputContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  min-width: 300px;
  border-radius: 7px;
  border: var(--borderUL);
  margin: 7px;
  margin-top: 0;
  padding: 7px 4px;
`
const PacklistNameInputLabel = styled(Label)`
  text-align: left;
`
const AddButton = styled.button`
  align-self: center;
  margin: 14px;
`
const StyledListContainer = styled(ListContainer)`
  border: none;
`
const TextSpan = styled.span`
  width: 100%;
  color: var(--font-color);
`
const HiddenInput = styled.input`
  display: none;
`
const ItemContainer = styled.label`
  display: flex;
  align-items: center;
`
const DeleteButton = styled.button`
  color: var(--deleteX);
  float: right;
  text-decoration: none;
  border: none;
  background: none;
  padding: 0;
`
const TemplateDropDown = styled(SelectTemplate)`
  grid-column: 1;
  grid-row: 8;
`
