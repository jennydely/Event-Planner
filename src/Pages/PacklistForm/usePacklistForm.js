import { useState } from 'react'
import usePacklists from '../Packlist/usePacklists'

export default function usePacklistForm() {
  const [packlistFormIsVisible, setPacklistFormIsVisible] = useState(false)
  const { updatePacklist, addPacklist } = usePacklists()

  const showPacklistForm = () => setPacklistFormIsVisible(true)

  const onPacklistSave = (packlist) => {
    addPacklist(packlist)
    setPacklistFormIsVisible(false)
  }
  const onPacklistSaveEdit = (packlist) => {
    updatePacklist(packlist)
    return true
  }
  const goPacklistBack = () =>
    window.location.reload() && setPacklistFormIsVisible(false)

  return {
    packlistFormIsVisible,
    showPacklistForm,
    onPacklistSave,
    goPacklistBack,
    onPacklistSaveEdit,
  }
}
