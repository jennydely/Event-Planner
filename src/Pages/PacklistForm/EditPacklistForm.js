import React from 'react'
import Footer from '../components/Footer'
import PacklistForm from './PacklistForm'

export default function EditPacklistForm() {
  return (
    <>
      <header>
        <h1>Create PackList</h1>
      </header>
      <main>
        <PacklistForm />
      </main>
      <Footer />
    </>
  )
}
