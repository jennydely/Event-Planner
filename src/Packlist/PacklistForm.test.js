import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import PacklistForm from './PacklistForm'
import 'jest-styled-components'

window.MutationObserver = require('mutation-observer')

describe('PacklistForm', () => {
  it('creates a new packlist with item', async () => {
    const onPacklistSave = jest.fn()
    const { getByText, getByPlaceholderText } = render(
      <PacklistForm onPacklistSave={onPacklistSave} />
    )

    fireEvent.input(getByPlaceholderText('PackList name'), {
      target: { value: 'Packlist' },
    })
    fireEvent.input(
      getByPlaceholderText('Item you need or task you have to do'),
      {
        target: { value: 'needed item' },
      }
    )
    fireEvent.click(getByText('Add'))
    expect(getByText('needed item')).toBeInTheDocument()

    fireEvent.submit(getByText('Save'))
    await waitFor(() => expect(onPacklistSave).toHaveBeenCalled())
  })

  it('displays a form to create packlist', async () => {
    const onPacklistSave = jest.fn()
    const { getByText, getByPlaceholderText } = render(
      <PacklistForm onPacklistSave={onPacklistSave} />
    )

    const nameInput = getByPlaceholderText('PackList name')
    expect(nameInput.value).toBe('')
    fireEvent.input(nameInput, { target: { value: 'My PacklistName' } })
    expect(nameInput.value).toBe('My PacklistName')

    fireEvent.click(getByText('Reset'))
    await waitFor(() => expect(nameInput.value).toBe(''))
  })

  it('renders correctly', () => {
    const onPacklistSave = jest.fn()
    const tree = render(<PacklistForm onPacklistSave={onPacklistSave} />)
    expect(tree).toMatchSnapshot()
  })
})
