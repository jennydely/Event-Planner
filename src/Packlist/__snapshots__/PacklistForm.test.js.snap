// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PacklistForm renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c4 {
  display: block;
  width: 100%;
  padding: 20px;
  border-radius: 4px;
  border: var(--border-darkgrey);
  margin-top: 0;
  padding: 4px;
  font-size: 112.5%;
  color: black;
}

.c2 {
  font-weight: 400;
  size: 112.5%;
  margin: 4px;
  text-align: right;
}

.c7 {
  display: block;
  width: 95%;
  border-radius: 7px;
  border: var(--border-darkgrey);
  margin: 7px;
  margin-top: 0;
  padding: 7px 4px;
}

.c0 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  min-width: 300px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  min-width: 300px;
  border-radius: 7px;
  border: var(--border-darkgrey);
  margin: 7px;
  margin-top: 0;
  padding: 7px 4px;
  padding-bottom: 20px;
}

.c3 {
  text-align: left;
}

.c6 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin: 14px;
}

.c8 {
  border: none;
}

.c5 {
  display: none;
}

.c9 {
  grid-column: 1;
  grid-row: 8;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  margin: 7px;
  margin-top: 30px;
}

.c10 {
  background-color: var(--darkred-75);
}

<body>
    <div>
      <form
        class="c0"
        data-testid="packlistform"
      >
        <div
          class="c1"
        >
          <label
            class="c2 c3"
            for="name"
          >
            Name of the new PackList:
          </label>
          <input
            class="c4"
            id="name"
            name="name"
            placeholder="PackList name"
          />
          <label
            class="c2"
            for="itemInput"
          >
            Create new item or task:
          </label>
          <input
            class="c4"
            id="itemInput"
            name="item"
            placeholder="Item you need or task you have to do"
          />
          <input
            class="c5"
            name="shouldHaveOneItem"
          />
          <button
            class="c6"
            type="button"
          >
            Add
          </button>
          <ul
            class="c7 c8"
          />
        </div>
        <div
          class="c9"
        >
          <button
            type="reset"
          >
            Reset
          </button>
          <button
            class="c10"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="PacklistForm__Form-sc-1jem4nf-0 eBgvOZ"
      data-testid="packlistform"
    >
      <div
        class="PacklistForm__FormInputContainer-sc-1jem4nf-1 cTCBWH"
      >
        <label
          class="Label-sc-4r18cf-0 PacklistForm__PacklistNameInputLabel-sc-1jem4nf-2 kTDDAh VRePq"
          for="name"
        >
          Name of the new PackList:
        </label>
        <input
          class="Input-sc-1qrq46f-0 hKSNO"
          id="name"
          name="name"
          placeholder="PackList name"
        />
        <label
          class="Label-sc-4r18cf-0 kTDDAh"
          for="itemInput"
        >
          Create new item or task:
        </label>
        <input
          class="Input-sc-1qrq46f-0 hKSNO"
          id="itemInput"
          name="item"
          placeholder="Item you need or task you have to do"
        />
        <input
          class="PacklistForm__HiddenInput-sc-1jem4nf-6 cHjHNk"
          name="shouldHaveOneItem"
        />
        <button
          class="PacklistForm__AddButton-sc-1jem4nf-3 eERlGT"
          type="button"
        >
          Add
        </button>
        <ul
          class="ListContainer-sc-106yz72-0 PacklistForm__ItemContainer-sc-1jem4nf-4 fNjykF dUWpou"
        />
      </div>
      <div
        class="PacklistForm__ButtonGroup-sc-1jem4nf-8 hqTPqe"
      >
        <button
          type="reset"
        >
          Reset
        </button>
        <button
          class="PacklistForm__SubmitButton-sc-1jem4nf-9 gaBTWT"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
